extensions/amp-form/0.1/amp-form.js

```diff
@@ -436,29 +448,75 @@ export class AmpForm {
    */
   handleXhrSubmit_(varSubsFields, trust) {
     this.setState_(FormState_.SUBMITTING);
-
-    const p = this.doVarSubs_(varSubsFields)
-        .then(() => {
-          this.triggerFormSubmitInAnalytics_('amp-form-submit');
-          this.actions_.trigger(
-              this.form_, 'submit', /* event */ null, trust);
-          // After variable substitution
-          const values = this.getFormAsObject_();
-          // At the form submitting state, we want to display any template
-          // messages with the submitting attribute.
-          this.renderTemplate_(values);
-        })
-        .then(() => this.doActionXhr_())
-        .then(response => this.handleXhrSubmitSuccess_(response),
-            error => {
-              return this.handleXhrSubmitFailure_(/** @type {!Error} */(error));
-            });
-
+    const varSubPromise = this.doVarSubs_(varSubsFields);
+    let p;
+    if (this.ssrTemplateHelper_.isSupported()) {
+      p = varSubPromise.then(() => {
+        this.actions_.trigger(
+            this.form_, 'submit', /* event */ null, trust);
+      }).then(() => {
+        return this.ssrTemplateHelper_.fetchAndRenderTemplate(
+            this.form_,
+            this.handleSsrTemplateSuccess_.bind(this),
+            this.handleSsrTemplateFailure_.bind(this)
+        );
+      });
+    } else {
+      p = varSubPromise
+          .then(() => {
+            this.handleFormSubmitting_(trust);
+          })
+          .then(() => this.doActionXhr_())
+          .then(response => this.handleXhrSubmitSuccess_(response),
+              error => {
+                return this.handleXhrSubmitFailure_(/** @type {!Error} */(error));
+              });
+    }
     if (getMode().test) {
       this.xhrSubmitPromise_ = p;
     }
   }
 
+  /**
+   * Handles viewer render template success.
+   * @param {!JsonObject} response
+   * @return {!Promise}
+   */
+  handleSsrTemplateSuccess_(response) {
+    this.setState_(FormState_.SUBMIT_SUCCESS);
+    return tryResolve(() => {
+      this.renderTemplate_(response.renderedHtml || {});
+    });
+  }
+
+  /**
+   * Handles viewer render template failure.
+   * @param {!JsonObject} error
+   */
+  handleSsrTemplateFailure_(error) {
+    this.triggerAction_(/* success */ false, error); // do we need this?
+    this.setState_(FormState_.SUBMIT_ERROR);
+    user().error(TAG, `Form submission failed: ${error}`);
+    return tryResolve(() => {
+      this.renderTemplate_(error || {});
+    });
+  }
+
+  /**
+   * Triggers the analytics and renders any template for submitting state.
+   * @param {ActionTrust} trust
+   */
+  handleFormSubmitting_(trust) {
```

Comment:
`submittingWithTrust_()`. Using "handle" as a prefix is confusing since it sounds like we're responding to an event or action.